File: leetcode-cpp/_30xx/_3169_count_days_without_meetings.cpp
#include <algorithm>
#include <iostream>

#include "../test/catch_amalgamated.hpp"
using namespace std;

class Solution {
   public:
    int countDays(int days, vector<vector<int>>& meetings) {
        int x = 1;
        std::sort(meetings.begin(), meetings.end(), [](const vector<int>& a, const vector<int>& b) {
            return a[0] < b[0];
        });
        // merge the meetings
        vector<vector<int>> merged{};
        for (auto meeting : meetings) {
            if (merged.size() == 0) {
                merged.push_back(meeting);
                continue;
            }
            auto& last = merged.back();
            if (last[1] < meeting[0]) {
                merged.push_back(meeting);
            } else {
                last[1] = std::max(last[1], meeting[1]);
            }
        }
        int meeting_count = 0;
        for (auto meeting : merged) {
            meeting_count += meeting[1] - meeting[0] + 1;
        }

        return days - meeting_count;
    }
};

TEST_CASE("test_case_1") {
    Solution s;
    auto meetings = vector<vector<int>>{{5, 7}, {1, 3}, {9, 10}};
    auto actual = s.countDays(10, meetings);
    CHECK(actual == 2);
}

TEST_CASE("test_case_2") {
    Solution s;
    auto meetings = vector<vector<int>>{{2, 4}, {1, 3}};
    auto actual = s.countDays(5, meetings);
    CHECK(actual == 1);
}

TEST_CASE("test_case_3") {
    Solution s;
    auto meetings = vector<vector<int>>{{1, 6}};
    auto actual = s.countDays(6, meetings);
    CHECK(actual == 0);
}

File: leetcode-python/_30xx/_3169_count_days_without_meetings.py
import unittest
from typing import List
import collections
#2025-24-03
class Solution:
    def countDays(self, days: int, meetings: List[List[int]]) -> int:
        meetings.sort() # sort in-place
        # merge meetings
        merged = []
        for [start, end] in meetings:
            if not merged:
                merged.append([start, end])
                continue
            last = merged[-1]
            if last[1] < start:
                merged.append(([start, end]))
            else:
                last[1] = max(end, last[1])

        meeting_count = 0
        for [start, end] in merged:
            meeting_count += end - start + 1
        return days - meeting_count



class MyTestCase(unittest.TestCase):

    def setUp(self) -> None:
        self.s = Solution()
    
    def test_case_1(self):
        actual = self.s.countDays(days = 10, meetings = [[5,7],[1,3],[9,10]])
        self.assertEqual(2, actual)
        
    def test_case_2(self):
        actual = self.s.countDays(days = 5, meetings = [[2,4],[1,3]])
        self.assertEqual(1, actual)

    def test_case_3(self):
        actual = self.s.countDays(days = 6, meetings = [[1,6]])
        self.assertEqual(0, actual)

    def test_case_4(self):
        actual = self.s.countDays(days = 10, meetings = [[5,7],[1,3],[9,10], [1,4]])
        self.assertEqual(1, actual)

    def test_case_5(self):
        actual = self.s.countDays(days = 8, meetings = [[2,3],[3,5],[8,8]])
        self.assertEqual(3, actual)

if __name__ == '__main__':
    unittest.main()

